src/common/stack.rs

```diff
@@ -1,88 +1,419 @@
+
+		let index = self.stack.len().checked_sub(1)?.checked_sub(depth)?;
+		debug_assert!(self.stack.len() > index, "guaranteed by previous line");
+		Some(&self.stack[index])
+	}
+
+	/// Return mutable reference to item `depth` distance away from top
+	///
+	/// Does not check whether depth is in range.
+	pub(crate) fn get_relative_to_top_mut_unchecked(&mut self, depth: usize) -> &mut T {
+		let offset = self.stack.len() - 1 - depth;
+		&mut self.stack[offset]
+	}
+
+	/// Swaps two elements in the stack.
+	///
+	/// # Arguments
+	///
+	/// * a - The index of the first element
+	/// * b - The index of the second element
+	///
+	/// # Panics
+	///
+	/// Panics if `a` or `b` are out of bound.
+	#[inline]
+	pub(crate) fn swap(&mut self, a: usize, b: usize) {
+		self.stack.swap(a, b)
+	}
+
+	/// Removes an element from the vector and returns it.
+	///
+	/// The removed element is replaced by the last element of the vector.
+	///
+	/// This does not preserve ordering, but is O(1).
+	///
+	/// # Panics
+	///
+	/// Panics if `index` is out of bounds.
+	pub fn swap_remove(&mut self, index: usize) -> T {
+		self.stack.swap_remove(index)
+	}
+
+	pub(crate) fn top(&self) -> Option<&T> {
+		self.stack.last()
+	}
+
+	pub(crate) fn top_mut(&mut self) -> Option<&mut T> {
+		self.stack.last_mut()
+	}
+
+	// Same as Vec::[truncate](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.truncate)
+	pub(crate) fn truncate(&mut self, new_size: usize) {
+		self.stack.truncate(new_size)
+	}
+
+	#[inline]
+	pub(crate) fn len(&self) -> usize {
+		self.stack.len()
 	}
-}
 
-#[cfg(feature = "std")]
-impl error::Error for Error {
-	fn description(&self) -> &str {
-		&self.0
+	pub(crate) fn is_empty(&self) -> bool {
+		self.stack.is_empty()
 	}
 }
 
-/// Stack with limit.
-#[derive(Debug)]
-pub struct StackWithLimit<T> where T: Clone {
-	/// Stack values.
-	values: Vec<T>,
-	/// Stack limit (maximal stack len).
-	limit: usize,
+// Why introduce the extra complexity of StackSizeLimit, StackSizeInitial, and StackSize?
+// We want to make the user to do the correct thing using the type checker.
+// Check out the excellent Rust API Guidelines (linked below) for suggestions
+// about type safety in rust.
+//
+// By introducing the new typed arguments, we turn:
+//
+// ```
+// pub fn new(initial_size: usize, limit: usize) -> StackWithLimit<T> { ... }
+// ```
+//
+// into:
+//
+// ```
+// pub fn new(initial_size: StackSizeInitial<T>, limit: StackSizeLimit<T>) -> StackWithLimit<T> { ... }
+// ```
+//
+// https://rust-lang-nursery.github.io/api-guidelines/type-safety.html#c-custom-type
+
+/// Type for communicating the size of some contigous container.
+/// Used for constructing both [`StackSizeLimit`] and [`StackSizeInitial`].
+#[derive(Eq, PartialEq, Hash, Debug)]
+pub struct StackSize<T> {
+	num_elements: usize,
+	// PhantomData is a zero-sized type for keeping track of T without rustc complaining.
+	phantom: PhantomData<*const T>,
```

Comment:
Why `*const`? Why not just `T`? I think that the dropck works differently for `PhantomData<*const T>` compared to `PhantomData<T>` but I think you should at least add a comment explaining the choice.

Reply:
The docs recommend `PhantomData<*const T>` for when your type doesn't need to drop any `T`s. Something about drop check analysis.

https://doc.rust-lang.org/std/marker/struct.PhantomData.html#ownership-and-the-drop-check

Reply:
You're right, it's dropck related. https://doc.rust-lang.org/std/marker/struct.PhantomData.html#ownership-and-the-drop-check

Added an explanatory comment.